File: SIMSmeCoreLib/__SIMSme__/ObjC/Helper/TryCatchFinally.m
//
//  TryCatchFinally.m
//  SIMSme
//
//  Created by RBU on 27/01/16.
//  Copyright © 2020 ginlo.net GmbH. All rights reserved.
//

#import "TryCatchFinally.h"

void tryCatchFinally(void(^tryBlock)(void), void(^catchBlock)(NSException *e), void(^finallyBlock)(void)) {
    @try {
        tryBlock();
    }
    @catch (NSException *exception) {
        catchBlock(exception);
    }
    @finally {
        finallyBlock();
    }
}

File: SIMSmeCoreLib/__SIMSme__/Shared/TryCatchFinally.swift
//
//  TryCatchFinally.swift
//  SIMSme
//
//  Created by RBU on 27/01/16.
//  Copyright © 2020 ginlo.net GmbH. All rights reserved.
//

import Foundation

public func tryC(_ try: @escaping () -> Void) -> TryCatchFinally {
    TryCatchFinally(`try`)
}

public class TryCatchFinally {
    let tryFunc: () -> Void
    var catchFunc: (NSException) -> Void = { _ in
    }

    var finallyFunc: () -> Void = {}

    init(_ try: @escaping () -> Void) {
        tryFunc = `try`
    }

    @discardableResult
    public func `catch`(_ catch: @escaping (NSException) -> Void) -> TryCatchFinally {
        // objc bridging needs NSException!, not NSException as we'd like to expose to clients.
        self.catchFunc = { e in
            `catch`(e)
        }
        return self
    }

    public func finally(_ finally: @escaping () -> Void) {
        self.finallyFunc = finally
    }

    deinit {
        tryCatchFinally(self.tryFunc, self.catchFunc, self.finallyFunc)
    }
}
